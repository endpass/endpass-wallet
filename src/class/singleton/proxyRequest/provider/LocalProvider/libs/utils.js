import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
