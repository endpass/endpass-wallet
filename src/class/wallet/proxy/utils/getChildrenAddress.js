export default function(hdWallet, offset, limit) {
  return [...Array(limit)].map((_, i) =>
    hdWallet
      .deriveChild(offset + i)
      .getWallet()
      .getChecksumAddressString(),
  );
}
