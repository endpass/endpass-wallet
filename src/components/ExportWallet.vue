<template>
  <div class="export-wallet app-page">
    <div class="section">
      <div class="container">
        <div class="card app-card">
          <div class="card-header">
            <h1 class="card-header-title">Export Wallet</h1>
          </div>
          <div class="card-content">
            <a @click="$router.go(-1)">&lt; Back</a>
            <p class="subtitle">Exporting account <strong>{{address}}</strong></p>
            <div class="columns">

              <div class="column is-one-third">
                <div class="menu">
                  <p class="menu-label">Export Type</p>
                  <ul class="menu-list">
                    <li>
                      <a @click="exportType = 'privateKey'"
                      :class="{'is-active':exportType==='privateKey'}">Private
                      Key</a>
                    </li>
                    <li>
                      <a @click="exportType = 'seedPhrase'"
                      :class="{'is-active':exportType==='seedPhrase'}">Seed Phrase</a>
                    </li>
                    <li>
                      <a @click="exportType = 'json'"
                      :class="{'is-active':exportType==='json'}">JSON
                      Keystore</a>
                    </li>
                  </ul>
                </div>
              </div>

              <div class="column">
                <div v-if="exportType==='privateKey'">
                  <export-to-private-key/>
                </div>
                <div v-else-if="exportType==='json'">
                  <export-to-json/>
                </div>
                <div v-else>
                  <p class="subtitle">This export type is not supported on
                  this device.</p>
                </div>
              </div>

            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import ExportToJson from '@/components/ExportToJson.vue'
import ExportToPrivateKey from '@/components/ExportToPrivateKey.vue'
import accounts from '@/mixins/accounts'

export default {
  data () {
    return {
      exportType: 'privateKey',
    }
  },
  components: {
    ExportToJson,
    ExportToPrivateKey
  },
  mixins: [accounts]
}
</script>

<style lang="scss">
.menu-list {
  a.is-active {
    background-color: $purple;
  }
}
</style>
