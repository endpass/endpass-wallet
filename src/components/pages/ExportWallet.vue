<template>
  <base-page class="export-wallet">
    <template slot="title">
      Export Wallet
    </template>
    <p class="subtitle">
      Exporting account <strong>{{ address }}</strong>
    </p>
    <div class="columns">
      <div class="column is-one-third">
        <div class="menu">
          <p class="menu-label">
            Export Type
          </p>
          <ul class="menu-list">
            <li>
              <a
                :class="{ 'is-active': exportType === 'privateKey' }"
                data-test="export-private-key-button"
                @click="exportType = 'privateKey'"
              >
                Private Key
              </a>
            </li>
            <li>
              <a
                :class="{ 'is-active': exportType === 'seedPhrase' }"
                data-test="export-seed-phrase-button"
                @click="exportType = 'seedPhrase'"
              >
                Seed Phrase
              </a>
            </li>
            <li>
              <a
                :class="{ 'is-active': exportType === 'json' }"
                data-test="export-json-button"
                @click="exportType = 'json'"
              >
                JSON Keystore
              </a>
            </li>
          </ul>
        </div>
      </div>

      <div class="column">
        <div v-if="exportType === 'privateKey'">
          <export-to-private-key />
        </div>
        <div v-else-if="exportType === 'json'">
          <export-to-json />
        </div>
        <div v-else>
          <p class="subtitle">
            This export type is not supported on this device.
          </p>
        </div>
      </div>
    </div>
  </base-page>
</template>

<script>
import { mapState } from 'vuex';
import BasePage from '@/components/pages/Base';
import ExportToJson from '@/components/ExportToJson';
import ExportToPrivateKey from '@/components/ExportToPrivateKey';
import privatePage from '@/mixins/privatePage';

export default {
  name: 'ExportWallet',
  data() {
    return {
      exportType: 'privateKey',
    };
  },
  computed: {
    ...mapState({
      address: state => state.accounts.address,
    }),
  },

  mixins: [privatePage],

  components: {
    BasePage,
    ExportToJson,
    ExportToPrivateKey,
  },
};
</script>

<style lang="scss">
.menu-list {
  a.is-active {
    background-color: $purple;
  }
}
</style>
