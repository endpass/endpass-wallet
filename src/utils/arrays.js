export const uniq = arr =>
  arr.reduce((acc, item) => (acc.includes(item) ? acc : acc.concat(item)), []);

export const mapArrayByProp = (arr, prop) =>
  arr.reduce((acc, item) => {
    const target = item[prop];

    if (target) {
      return Object.assign(acc, {
        [target]: item,
      });
    }

    return acc;
  }, {});

export default {
  uniq,
  mapArrayByProp,
};
