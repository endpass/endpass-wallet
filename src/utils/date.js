import dayjs from 'dayjs';
import relativeTime from 'dayjs/plugin/relativeTime';

/**
 * Setting up dayjs globally
 */
dayjs.extend(relativeTime);

export const formateDate = date => dayjs(date).format('YYYY-MM-DD H:mm');

export const fromNow = date => dayjs(date).fromNow();

export default {
  formateDate,
  fromNow,
};
