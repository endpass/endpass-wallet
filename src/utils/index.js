import store from '../store';

export { default as http } from './http';

export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());

export const getInitializedValueFromStore = (module, field) => {
  return new Promise(resolve => {
    const value = module[field];

    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
};
