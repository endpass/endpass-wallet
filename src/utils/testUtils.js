import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
