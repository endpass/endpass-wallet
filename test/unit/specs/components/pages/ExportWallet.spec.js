import { shallow } from '@vue/test-utils';
import Vuex from 'vuex';
import { generateStubs } from '@/utils/testUtils';

import ExportWallet from '@/components/pages/ExportWallet';

describe('ExportWallet page', () => {
  let wrapper;
  beforeEach(() => {
    const store = new Vuex.Store({
      state: {
        accounts: {},
      },
    });
    wrapper = shallow(ExportWallet, {
      store,
      stubs: generateStubs(ExportWallet),
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('export-wallet');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should display correct components depending on export type', () => {
      wrapper.setData({
        exportType: 'privateKey',
      });
      expect(wrapper.find('export-to-private-key').exists()).toBe(true);
      expect(wrapper.find('export-to-json').exists()).toBe(false);
      wrapper.setData({
        exportType: 'json',
      });
      expect(wrapper.find('export-to-json').exists()).toBe(true);
      expect(wrapper.find('export-to-private-key').exists()).toBe(false);
      wrapper.setData({
        exportType: 'seedPhrase',
      });
      expect(wrapper.find('export-to-json').exists()).toBe(false);
      expect(wrapper.find('export-to-private-key').exists()).toBe(false);
      //type is unsupportable message
      expect(wrapper.find('.subtitle').exists()).toBe(true);
    });
  });
});
