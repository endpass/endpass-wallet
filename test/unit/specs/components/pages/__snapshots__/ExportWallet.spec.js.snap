// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportWallet page render should render initial state of the component 1`] = `
<base-page
  class="export-wallet"
>
  Export Wallet 
  <p
    class="subtitle"
  >
    
    Exporting account 
    <strong>
      
    </strong>
  </p>
   
  <div
    class="columns"
  >
    <div
      class="column is-one-third"
    >
      <div
        class="menu"
      >
        <p
          class="menu-label"
        >
          Export Type
        </p>
         
        <ul
          class="menu-list"
        >
          <li>
            <a
              class="is-active"
            >
              
              Private Key
            
            </a>
          </li>
           
          <li>
            <a
              class=""
            >
              
              Seed Phrase
            
            </a>
          </li>
           
          <li>
            <a
              class=""
            >
              
              JSON Keystore
            
            </a>
          </li>
        </ul>
      </div>
    </div>
     
    <div
      class="column"
    >
      <div>
        <export-to-private-key />
      </div>
    </div>
  </div>
</base-page>
`;
